lib/std/fs/file.zig

```diff
@@ -385,6 +385,274 @@ pub const File = struct {
+                const shift = @as(u3, @enumToInt(class)) * 3;
+                if (permissions.read) |r| {
+                    if (r) {
+                        self.mode |= @as(Mode, 0o4) << shift;
+                    } else {
+                        self.mode &= ~(@as(Mode, 0o4) << shift);
+                    }
+                }
+                if (permissions.write) |w| {
+                    if (w) {
+                        self.mode |= @as(Mode, 0o2) << shift;
+                    } else {
+                        self.mode &= ~(@as(Mode, 0o2) << shift);
+                    }
+                }
+                if (permissions.execute) |x| {
+                    if (x) {
+                        self.mode |= @as(Mode, 0o1) << shift;
+                    } else {
+                        self.mode &= ~(@as(Mode, 0o1) << shift);
+                    }
+                }
+            }
+
+            /// Returns a `Permissions` struct representing the permissions from the passed mode.
+            /// This method is only available on Unix platforms.
+            pub fn unixNew(new_mode: Mode) Self {
+                return Self{
+                    .mode = new_mode,
+                };
+            }
+        },
+    };
+
+    pub const SetPermissionsError = ChmodError;
+
+    /// Sets permissions according to the provided `Permissions` struct.
+    /// This method is *NOT* available on WASI
+    pub fn setPermissions(self: File, permissions: Permissions) SetPermissionsError!void {
+        switch (builtin.os.tag) {
+            .windows => {
+                var io_status_block: windows.IO_STATUS_BLOCK = undefined;
+                var info = windows.FILE_BASIC_INFORMATION{
+                    .CreationTime = 0,
+                    .LastAccessTime = 0,
+                    .LastWriteTime = 0,
+                    .ChangeTime = 0,
+                    .FileAttributes = permissions.attributes,
+                };
+                const rc = windows.ntdll.NtSetInformationFile(
+                    self.handle,
+                    &io_status_block,
+                    &info,
+                    @sizeOf(windows.FILE_BASIC_INFORMATION),
+                    .FileBasicInformation,
+                );
+                switch (rc) {
+                    .SUCCESS => return,
+                    .INVALID_HANDLE => unreachable,
+                    .ACCESS_DENIED => return error.AccessDenied,
+                    else => return windows.unexpectedStatus(rc),
+                }
+            },
+            .wasi => @compileError("Unsupported OS"), // TODO Waiting for wasi-filesystem to support chmod
+            else => {
+                try self.chmod(permissions.mode);
+            },
+        }
+    }
+
+    /// Cross-platform representation of file metadata.
+    /// Can be obtained with `File.metadata()`
+    pub const Metadata = switch (builtin.os.tag) {
+        .windows => struct {
+            attributes: windows.DWORD,
```

Comment:
I wonder if we should prefix all platform specific fields with a platform identifier?  So in this case `windowsAttributes`?  The reason I suggest this is to make it obvious what parts of the API are platform agnostic and which aren't.   For example,
```zig
const m = try file.metadata();
std.log.info("file attributes={d}", .{m.attributes});
```
This code snippet looks like it's platform agnostic but it's not.  If the developer who wrote this only compiles it for Windows then they won't know the mistake.  This would become much more obvious if it looked like this:

```zig
const m = try file.metadata();
std.log.info("file attributes={d}", .{m.windowsAttributes});
```

Now it's alot more obvious that this field is specific to Windows and if they want their code to work elsewhere to surround it with a `builtin.os.tag` check.  We also use this strategy for platform specific functions like you did with `unixHas` etc.

Reply:
I'm starting to rethink a lot of this really, and one idea I keep coming back do is having `File.Metadata`, which consistently has the same functions, and has a field `inner`, which contains the actual platform-specific implementation (`MetadataLinux` etc), and has methods which aren't shared across platforms. This would help discourage direct use of member fields as well as prefixing them with the platform name, and additionally keep documentation for `File.Metadata` exclusively to cross-platform methods.

Reply:
Yeah that's an improvement IMO.

Reply:
This should now be resolved by the pushed changes, though I haven't yet added any platform-specific operations :)